### DISCUSSION TITLE
Why my model doesn't learn?

### DISCUSSION THREAD

**Discussion by rezareza007** [coqui-ai/TTS#1847] [General Q&A]
I use dataset of about 2000 noiseless records with the following configuration. I'm using graphemes instead of phonemes because of the peculiarities of the language. The records range from long sentences to parts of sentences to one-word exclamations. 

>  {
    "output_path":  "./runs", 
    "logger_uri": null,
    "run_name": "run_new_1",
    "project_name": null,
   // "min_seq_len": 6,
   // "max_seq_len": 250,
    "run_description": "Coqui trainer run.",
    "print_step": 25,
    "plot_step": 100,
    "model_param_stats": false,
    "wandb_entity": null,
    "dashboard_logger": "tensorboard",
    "log_model_step": null,
    "save_step": 10000,
    "save_n_checkpoints": 5,
    "save_checkpoints": true,
    "save_all_best": false,
    "save_best_after": 10000,
    "target_loss": null,
    "print_eval": true,
    "test_delay_epochs": -1,
    "run_eval": true,
    "distributed_backend": "nccl",
    "mixed_precision": false,
    "epochs": 1000,
    "batch_size": 40    ,
    "eval_batch_size": 16,
    "grad_clip": 5.0,
    "scheduler_after_epoch": true,
    "lr": 0.0001,
    "optimizer": "RAdam",
   
    "optimizer_params": {
        "betas": [
            0.9,
            0.998
        ],
        "weight_decay": 1e-06
    },
    "lr_scheduler": "NoamLR",
    "lr_scheduler_params": {
        "warmup_steps": 4000
    },
    "use_grad_scaler": false,
    "cudnn_enable": true,
    "cudnn_deterministic": false,
    "cudnn_benchmark": false,
    "training_seed": 54321,
    "model": "tacotron2",
    "num_loader_workers": 4,
    "num_eval_loader_workers": 4,
    "use_noise_augment": false,
    "audio": {
        "fft_size": 1024,
        "win_length": 1024,
        "hop_length": 256,
        "frame_shift_ms": null,
        "frame_length_ms": null,
        "stft_pad_mode": "reflect",
        "sample_rate": 22050,
        "resample": false,
        "preemphasis": 0.0,
        "ref_level_db": 20,
        "do_sound_norm": false,
        "log_func": "np.log",
        "do_trim_silence": true,
        "trim_db": 60.0,
        "do_rms_norm": false,
        "db_level": null,
        "power": 1.5,
        "griffin_lim_iters": 60,
        "num_mels": 80,
        "mel_fmin": 0.0,
        "mel_fmax": 8000,
        "spec_gain": 1.0,
        "do_amp_to_db_linear": true,
        "do_amp_to_db_mel": true,
        "pitch_fmax": 640.0,
        "pitch_fmin": 0.0,
        "signal_norm": false,
        "min_level_db": -100,
        "symmetric_norm": true,
        "max_norm": 4.0,
        "clip_norm": true,
        "stats_path":  "data/stats.npy"
    },
    "use_phonemes": false,
    // "phonemizer": "espeak",
    // "phoneme_language": "fa",
    "compute_input_seq_cache": false,
    "text_cleaner": null,
    "enable_eos_bos_chars": false,
    "test_sentences_file": null,
    "characters": {
        "characters_class": "TTS.tts.utils.text.characters.Graphemes",
        "vocab_dict": null, // {0: "a", 1: "b", 2: "c", 3: "d", 4: "e", 5: "f", 6: "h", 7: "i", 8: "j", 9: "k", 10: "l", 11: "m", 12: "n", 13: "o", 14: "p", 15: "q", 16: "r", 17: "s", 18: "t", 19: "u", 20: "v", 21: "w", 22: "x", 23: "y", 24: "z", 25: "æ", 26: "ç", 27: "ð", 28: "ø", 29: "ħ", 30: "ŋ", 31: "œ", 32: "ǀ", 33: "ǁ", 34: "ǂ", 35: "ǃ", 36: "ɐ", 37: "ɑ", 38: "ɒ", 39: "ɓ", 40: "ɔ", 41: "ɕ", 42: "ɖ", 43: "ɗ", 44: "ɘ", 45: "ə", 46: "ɚ", 47: "ɛ", 48: "ɜ", 49: "ɞ", 50: "ɟ", 51: "ɠ", 52: "ɡ", 53: "ɢ", 54: "ɣ", 55: "ɤ", 56: "ɥ", 57: "ɦ", 58: "ɧ", 59: "ɨ", 60: "ɪ", 61: "ɫ", 62: "ɬ", 63: "ɭ", 64: "ɮ", 65: "ɯ", 66: "ɰ", 67: "ɱ", 68: "ɲ", 69: "ɳ", 70: "ɴ", 71: "ɵ", 72: "ɶ", 73: "ɸ", 74: "ɹ", 75: "ɺ", 76: "ɻ", 77: "ɽ", 78: "ɾ", 79: "ʀ", 80: "ʁ", 81: "ʂ", 82: "ʃ", 83: "ʄ", 84: "ʈ", 85: "ʉ", 86: "ʊ", 87: "ʋ", 88: "ʌ", 89: "ʍ", 90: "ʎ", 91: "ʏ", 92: "ʐ", 93: "ʑ", 94: "ʒ", 95: "ʔ", 96: "ʕ", 97: "ʘ", 98: "ʙ", 99: "ʛ", 100: "ʜ", 101: "ʝ", 102: "ʟ", 103: "ʡ", 104: "ʢ", 105: "ˈ", 106: "ˌ", 107: "ː", 108: "ˑ", 109: "˞", 110: "β", 111: "θ", 112: "χ", 113: "ᵻ", 114: "ⱱ"}, // null,
        "pad": "<PAD>",
        "eos": "<EOS>",
        "bos": "<BOS>",
        "blank": "<BLNK>",
        "characters": " پشَعُأثغزچوآٌهقلًانٍختسجضدکإحطةریصۀگظمؤيِّفءبژذئ",
        "punctuations": "!'()-.:؛؟،",
        "phonemes": null,
        "is_unique": false,
        "is_sorted": true
    },
    "add_blank": false,
    "batch_group_size": 0,
    "loss_masking": true,
    "sort_by_audio_len": false,
    "min_audio_len": 1,
    "max_audio_len": Infinity,
    "min_text_len": 1,
    "max_text_len": 150,
    "compute_f0": false,
    "compute_linear_spec": false,
    "precompute_num_workers": 0,
    "start_by_longest": false,
    "datasets": [
        {
            "name": "ljspeech",
            "path": "./",
            "meta_file_train": "metadata_.csv",
            "ignored_speakers": null,
            "language": "",
            "meta_file_val": "",
            "meta_file_attn_mask": ""
        }
    ],
    "test_sentences": [
        "سلام. من رضا  هستم.",
        "زندگی منشوریست در حرکت دوار.",
        "چرا اینطوریه؟",
    ],
    "eval_split_max_size": null,
    "eval_split_size": 0.01,
    "use_speaker_weighted_sampler": false,
    "speaker_weighted_sampler_alpha": 1.0,
    "use_language_weighted_sampler": false,
    "language_weighted_sampler_alpha": 1.0,
    "use_gst": false,
    "gst": null,
    "gst_style_input": null,
    "num_speakers": 1,
    "num_chars": 67,
    "r": 2,
    "gradual_training": null,
    "memory_size": -1,
    "prenet_type": "original",
    "prenet_dropout": true,
    "prenet_dropout_at_inference": false,
    "stopnet": true,
    "separate_stopnet": true,
    "stopnet_pos_weight": 10.0,
    "max_decoder_steps": 1000,
    "encoder_in_features": 512,
    "decoder_in_features": 512,
    "decoder_output_dim": 80,
    "out_channels": 80,
    "attention_type": "dynamic_convolution", //"bidirectional_decoder", // 
    "attention_heads": null,
    "attention_norm": "sigmoid",
    "attention_win": false,
    "windowing": false,
    "use_forward_attn": false,
    "forward_attn_mask": false,
    "transition_agent": false,
    "location_attn": true,
    "bidirectional_decoder": false,
    "double_decoder_consistency": false,
    "ddc_r": 6,
    "speakers_file": null,
    "use_speaker_embedding": false,
    "speaker_embedding_dim": 512,
    "use_d_vector_file": false,
    "d_vector_file": false,
    "d_vector_dim": null,
    "seq_len_norm": false,
    "decoder_loss_alpha": 0.25,
    "postnet_loss_alpha": 0.25,
    "postnet_diff_spec_alpha": 0,
    "decoder_diff_spec_alpha": 0,
    "decoder_ssim_alpha": 0,
    "postnet_ssim_alpha": 0,
    "ga_alpha": 0.0,
    "github_branch": "inside_docker"
}

![image](https://user-images.githubusercontent.com/72138796/183622471-4fa6ef9f-4f56-4868-a841-e5fc3d864918.png)
![image](https://user-images.githubusercontent.com/72138796/183622592-2794e295-df5b-4484-bf31-0c2543577549.png)

First of all, here is the latest step in the training:

   --> STEP: 55/57 -- GLOBAL_STEP: 54775
     | > decoder_loss: 0.71597  (0.66532)
     | > postnet_loss: 0.61301  (0.57915)
     | > stopnet_loss: 0.06795  (0.10856)
     | > loss: 0.40020  (0.41968)
     | > align_error: 0.95695  (0.94082)
     | > grad_norm: 1.79131  (1.41701)
     | > current_lr: 0.00002
     | > step_time: 2.26630  (0.65488)
     | > loader_time: 0.00600  (0.01174)

but tensorboard shows only 800 steps ?? The generated voice is garbled with a long 'aaaaaaa'
sound at the end.
My main question is if the problem is caused by the rather small size of my dataset or is there something else I'm doing wrong?
I'm curating a larger, better datase but for now, would a two speaker dataset help improve training?
How about other attention mechanisms like bidirectional_decoder help?

**Comment by rezareza007** [comment]
Bump. Help please!

**Comment by KyuubiYoru** [comment]
I think your dataset is too small to train Tacotron2, but you could try to improve your audio settings using "TTS/Notesbooks/dataset_analysis/CheckSpectrograms.ipynb".

800 Steps is not enough training to see anything, you need a few thousand to slowly see a training progress.

Maybe the VITS model works better.
